File: src/translators/while_to_goto.ml
(** Translate WHILE programs to GOTO programs
    
    This demonstrates the equivalence between WHILE and GOTO (FSK §10.4).
    
    Key translations:
    1. Sequential composition is already compatible
    2. WHILE c DO P END becomes:
       L_start: IF NOT c THEN GOTO L_end
       P (translated)
       GOTO L_start
       L_end: SKIP
*)

(** Convert WHILE expression to GOTO expression *)
let rec translate_expr (e : While_lang.Ast.expr) : Goto.Ast.expr =
  match e with
  | While_lang.Ast.Const n -> Goto.Ast.Const n
  | While_lang.Ast.Var x -> Goto.Ast.Var x
  | While_lang.Ast.Add (e1, e2) -> Goto.Ast.Add (translate_expr e1, translate_expr e2)
  | While_lang.Ast.Sub (e1, e2) -> Goto.Ast.Sub (translate_expr e1, translate_expr e2)
  | While_lang.Ast.Mul (e1, e2) -> Goto.Ast.Mul (translate_expr e1, translate_expr e2)

(** Convert comparison operator *)
let translate_comp_op (op : While_lang.Ast.comp_op) : Goto.Ast.comp_op =
  match op with
  | While_lang.Ast.Eq -> Goto.Ast.Eq
  | While_lang.Ast.Neq -> Goto.Ast.Neq
  | While_lang.Ast.Lt -> Goto.Ast.Lt
  | While_lang.Ast.Gt -> Goto.Ast.Gt
  | While_lang.Ast.Leq -> Goto.Ast.Leq
  | While_lang.Ast.Geq -> Goto.Ast.Geq

(** Negate a comparison operator *)
let negate_comp_op (op : Goto.Ast.comp_op) : Goto.Ast.comp_op =
  match op with
  | Goto.Ast.Eq -> Goto.Ast.Neq
  | Goto.Ast.Neq -> Goto.Ast.Eq
  | Goto.Ast.Lt -> Goto.Ast.Geq
  | Goto.Ast.Gt -> Goto.Ast.Leq
  | Goto.Ast.Leq -> Goto.Ast.Gt
  | Goto.Ast.Geq -> Goto.Ast.Lt

(** Translate a condition into a series of GOTO instructions
    that jump to label_true if condition is true, label_false otherwise *)
let rec translate_condition 
    (c : While_lang.Ast.condition) 
    (label_true : string) 
    (label_false : string) : Goto.Ast.labeled_instr list =
  match c with
  | While_lang.Ast.Compare (e1, op, e2) ->
      let e1' = translate_expr e1 in
      let e2' = translate_expr e2 in
      let op' = translate_comp_op op in
      [
        { Goto.Ast.label = None; instr = Goto.Ast.IfGoto (e1', op', e2', label_true) };
        { Goto.Ast.label = None; instr = Goto.Ast.Goto label_false }
      ]
  
  | While_lang.Ast.Not c ->
      (* Swap true and false labels *)
      translate_condition c label_false label_true
  
  | While_lang.Ast.And (c1, c2) ->
      let label_c2 = Common.Utils.fresh_label "and_" in
      (* If c1 is false, jump to label_false; if true, check c2 *)
      (translate_condition c1 label_c2 label_false) @
      [{ Goto.Ast.label = Some label_c2; instr = Goto.Ast.Skip }] @
      (translate_condition c2 label_true label_false)
  
  | While_lang.Ast.Or (c1, c2) ->
      let label_c2 = Common.Utils.fresh_label "or_" in
      (* If c1 is true, jump to label_true; if false, check c2 *)
      (translate_condition c1 label_true label_c2) @
      [{ Goto.Ast.label = Some label_c2; instr = Goto.Ast.Skip }] @
      (translate_condition c2 label_true label_false)

(** Translate WHILE program to GOTO instructions *)
let rec translate (p : While_lang.Ast.program) : Goto.Ast.labeled_instr list =
  match p with
  | While_lang.Ast.Skip ->
      [{ Goto.Ast.label = None; instr = Goto.Ast.Skip }]
  
  | While_lang.Ast.Assign (x, e) ->
      [{ Goto.Ast.label = None; instr = Goto.Ast.Assign (x, translate_expr e) }]
  
  | While_lang.Ast.Seq (p1, p2) ->
      (translate p1) @ (translate p2)
  
  | While_lang.Ast.While (cond, body) ->
      let label_start = Common.Utils.fresh_label "while_start_" in
      let label_body = Common.Utils.fresh_label "while_body_" in
      let label_end = Common.Utils.fresh_label "while_end_" in
      
      (* L_start: check condition *)
      [{ Goto.Ast.label = Some label_start; instr = Goto.Ast.Skip }] @
      (translate_condition cond label_body label_end) @
      (* L_body: execute body *)
      [{ Goto.Ast.label = Some label_body; instr = Goto.Ast.Skip }] @
      (translate body) @
      (* Jump back to start *)
      [{ Goto.Ast.label = None; instr = Goto.Ast.Goto label_start }] @
      (* L_end: exit *)
      [{ Goto.Ast.label = Some label_end; instr = Goto.Ast.Skip }]

(** Translate and add HALT at the end *)
let translate_program (p : While_lang.Ast.program) : Goto.Ast.program =
  Common.Utils.reset_fresh_labels ();
  let instrs = translate p in
  instrs @ [{ Goto.Ast.label = None; instr = Goto.Ast.Halt }]

File: src/translators/whileToGoto.ts
import type { Program as WhileProgram, Statement as WhileStatement, Expression as WhileExpression, Condition as WhileCondition } from "../while/ast";
import type { Program as GotoProgram, Instruction, Statement as GotoStatement, Expression as GotoExpression, Condition as GotoCondition } from "../goto/ast";

// WHILE cond DO body END  =>  M_i: IF NOT cond THEN GOTO M_j; body; GOTO M_i; M_j: ...
export class WhileToGotoTranslator {
    private labelCounter = 0;
    private nextFreshVarIndex = 0;

    private findMaxVariableIndex(program: WhileProgram): number {
        let maxIndex = -1;

        const checkExpr = (expr: WhileExpression): void => {
            if (expr.type === "variable") {
                const match = expr.name.match(/^x(\d+)$/);
                if (match?.[1]) maxIndex = Math.max(maxIndex, parseInt(match[1], 10));
            } else if (expr.type === "binaryOp") {
                checkExpr(expr.left);
                checkExpr(expr.right);
            }
        };

        const checkStmt = (stmt: WhileStatement): void => {
            if (stmt.type === "assignment") {
                const match = stmt.variable.match(/^x(\d+)$/);
                if (match?.[1]) maxIndex = Math.max(maxIndex, parseInt(match[1], 10));
                checkExpr(stmt.value);
            } else if (stmt.type === "while") {
                checkExpr(stmt.condition.left);
                checkExpr(stmt.condition.right);
                stmt.body.forEach(checkStmt);
            } else if (stmt.type === "if") {
                checkExpr(stmt.condition.left);
                checkExpr(stmt.condition.right);
                stmt.thenBody.forEach(checkStmt);
                stmt.elseBody?.forEach(checkStmt);
            }
        };

        program.statements.forEach(checkStmt);
        return maxIndex;
    }

    private freshLabel(): string {
        return `M${this.labelCounter++}`;
    }

    private freshVar(): string {
        return `x${this.nextFreshVarIndex++}`;
    }

    public translate(program: WhileProgram): GotoProgram {
        this.labelCounter = 0;
        this.nextFreshVarIndex = this.findMaxVariableIndex(program) + 1;
        
        const instructions = this.translateStatements(program.statements);
        instructions.push({ statement: { type: "halt" } });
        
        return { type: "program", instructions };
    }

    private translateStatements(statements: WhileStatement[]): Instruction[] {
        const result: Instruction[] = [];
        for (const stmt of statements) {
            result.push(...this.translateStatement(stmt));
        }
        return result;
    }

    private createNoop(): GotoStatement {
        return {
            type: "assignment",
            variable: this.freshVar(),
            value: { type: "number", value: 0 }
        };
    }

    private translateStatement(stmt: WhileStatement): Instruction[] {
        if (stmt.type === "assignment") {
            return [{
                statement: {
                    type: "assignment",
                    variable: stmt.variable,
                    value: this.translateExpression(stmt.value)
                }
            }];
        }

        if (stmt.type === "while") {
            const labelStart = this.freshLabel();
            const labelEnd = this.freshLabel();

            return [
                {
                    label: labelStart,
                    statement: {
                        type: "if_goto",
                        condition: this.negateCondition(this.translateCondition(stmt.condition)),
                        label: labelEnd
                    }
                },
                ...this.translateStatements(stmt.body),
                { statement: { type: "goto", label: labelStart } },
                { label: labelEnd, statement: this.createNoop() }
            ];
        }

        // IF statement
        const labelElse = this.freshLabel();
        const labelEnd = this.freshLabel();
        const instructions: Instruction[] = [];

        if (stmt.elseBody && stmt.elseBody.length > 0) {
            instructions.push({
                statement: {
                    type: "if_goto",
                    condition: this.negateCondition(this.translateCondition(stmt.condition)),
                    label: labelElse
                }
            });
            instructions.push(...this.translateStatements(stmt.thenBody));
            instructions.push({ statement: { type: "goto", label: labelEnd } });

            const elseInstructions = this.translateStatements(stmt.elseBody);
            if (elseInstructions.length > 0) {
                elseInstructions[0] = { label: labelElse, statement: elseInstructions[0]!.statement };
                instructions.push(...elseInstructions);
            } else {
                instructions.push({ label: labelElse, statement: this.createNoop() });
            }
            instructions.push({ label: labelEnd, statement: this.createNoop() });
        } else {
            instructions.push({
                statement: {
                    type: "if_goto",
                    condition: this.negateCondition(this.translateCondition(stmt.condition)),
                    label: labelEnd
                }
            });
            instructions.push(...this.translateStatements(stmt.thenBody));
            instructions.push({ label: labelEnd, statement: this.createNoop() });
        }

        return instructions;
    }

    private negateCondition(cond: GotoCondition): GotoCondition {
        const negMap: Record<string, GotoCondition["operator"]> = {
            "=": "!=", "!=": "=", "<": ">=", ">": "<=", "<=": ">", ">=": "<"
        };
        return { type: "condition", operator: negMap[cond.operator]!, left: cond.left, right: cond.right };
    }

    private translateCondition(cond: WhileCondition): GotoCondition {
        return {
            type: "condition",
            operator: cond.operator,
            left: this.translateExpression(cond.left),
            right: this.translateExpression(cond.right)
        };
    }

    private translateExpression(expr: WhileExpression): GotoExpression {
        switch (expr.type) {
            case "number": return { type: "number", value: expr.value };
            case "variable": return { type: "variable", name: expr.name };
            case "binaryOp": return {
                type: "binaryOp",
                operator: expr.operator,
                left: this.translateExpression(expr.left),
                right: this.translateExpression(expr.right)
            };
        }
    }
}
